akka-http/src/main/scala/akka/http/javadsl/server/Directives.scala

```diff
@@ -26,4 +28,96 @@ object Directives extends AllDirectives {
 
   @varargs override def getFromBrowseableDirectories(directories: String*): Route =
     super.getFromBrowseableDirectories(directories: _*)
+
+  /**
+   * Partially applying functions for the Java World!
+   * @param f the function to partially apply
+   * @param a the first parameter to partially apply
+   * @tparam A the type of the applied parameter
+   * @tparam B the type of the second parameter
+   * @tparam R the type of the return
+   * @return the function partially applied
+   */
+  def partial[A, B, R](f: BiFunction[A, B, R], a: A): Function[B, R] = {
+    new Function[B, R] {
+      override def apply(b: B): R = f.apply(a, b)
+    }
+  }
+
+  /**
+   * Composes two 0-argument directives which share the same inner Route. This is equivalent to
+   * first(inner).orElse(second(inner))
+   * @param first the first 0-argument directive
+   * @param second the second 0-argument directive
+   * @param inner the inner route Producer
+   * @return the resulting route
+   */
+  def anyOf(first: Function[Supplier[Route], Route], second: Function[Supplier[Route], Route], inner: Supplier[Route]): Route = {
+    first.apply(inner).orElse(second.apply(inner))
+  }
+
+  /**
+   * Composes two 1-argument directives (directives whose inner route takes a parameter) which share the same inner route.
+   * This is the same as first(inner).orElse(second(inner)).
+   * This can be used also for directives that take an additional parameter if the directive is partially applied, ie. parameter
+   * @param first the first 1-argument directive
+   * @param second the second 1-argument directive
+   * @param inner the inner route that takes one argument
+   * @tparam A the type of the parameter the directives extract and the inner route takes
+   * @return the resulting route
+   */
+  def anyOf[A](first: Function[Function[A, Route], Route], second: Function[Function[A, Route], Route], inner: Function[A, Route]): Route = {
+    first.apply(inner).orElse(second.apply(inner))
+  }
+
+  /**
+   * Chains two 0-argument directives together
+   * @param first the 0-argument directive (outer one)
+   * @param second the 0-argument directive (inner one)
+   * @param inner the inner route function
+   * @return the resulting route
+   */
+  def allOf(first: Function[Supplier[Route], Route], second: Function[Supplier[Route], Route], inner: Supplier[Route]): Route = {
+    first.apply(new Supplier[Route] {
+      override def get(): Route =
+        second.apply(inner)
+    })
+  }
```

Comment:
Will be pretty decent to write once we can drop 2.11, with the lambda interop... ;)